ggshield/output/text/text_output_handler.py

```diff
@@ -26,49 +26,79 @@ class TextOutputHandler(OutputHandler):
-        else:
-            has_results = False
-            if scan.scans:
-                has_results = any(x.has_results for x in scan.scans)
+        scan_buf.write(self.process_scan_results(scan))
+
+        if scan.known_secrets_count > 0:
+            scan_buf.write(
+                f"\nWarning: {scan.known_secrets_count} {pluralize('secret', scan.known_secrets_count)} ignored "
+                f"because {pluralize('it is', scan.known_secrets_count, 'they are')} already known by your "
+                f"GitGuardian dashboard and you used the `--ignore-known-secrets` option.\n"
+            )
+
+            if self.verbose:
+                scan_buf.write(self.process_scan_results(scan, True))
+            else:
+                scan_buf.write("Use `--verbose` for more details.\n")
+
+        return scan_buf.getvalue()
+
+    def process_scan_results(
+        self, scan: ScanCollection, show_only_known_secrets: bool = False
+    ) -> str:
+        results_buf = StringIO()
+        if scan.results:
+            current_result_buf = StringIO()
+            for result in scan.results.results:
+                current_result_buf.write(
+                    self.process_result(result, show_only_known_secrets)
+                )
+            current_result_string = current_result_buf.getvalue()
+
+            # We want to show header in the verbose mode for new secrets or when at least one result is not empty
+            if scan.optional_header and (
+                current_result_string or (self.verbose and not show_only_known_secrets)
+            ):
+                results_buf.write(scan.optional_header)
 
-            if top and not has_results:
-                scan_buf.write(no_leak_message())
+            results_buf.write(current_result_string)
 
         if scan.scans:
             for sub_scan in scan.scans:
-                inner_scan_str = self._process_scan_impl(sub_scan, top=False)
-                scan_buf.write(inner_scan_str)
+                inner_scan_str = self.process_scan_results(
+                    sub_scan, show_only_known_secrets
+                )
+                results_buf.write(inner_scan_str)
 
-        return scan_buf.getvalue()
+        return results_buf.getvalue()
 
-    def process_result(self, result: Result) -> str:
+    def process_result(
+        self, result: Result, show_only_known_secrets: bool = False
+    ) -> str:
         """
         Build readable message on the found incidents.
 
         :param result: The result from scanning API
         :param nb_lines: The number of lines to display before and after a secret in the
         patch
         :param show_secrets: Option to show secrets value
+        :param show_only_known_secrets: If True, display only known secrets, and only new secrets otherwise
         :return: The formatted message to display
         """
         result_buf = StringIO()
+        result = deepcopy(result)
```

Comment:
I needed to do it, because we change content, policy breaks and matches in this function. It worked before since it was always executing only once. In future we could rework functions such that they do not change what is in the result.

Reply:
That's a good point. Can you add it as a comment in the code? It's going to be lost if it's only in the PR.

Also, from what I see we only change policy breaks, no? If that's the case then maybe we can restrict the copy to policy breaks by deepcopying only on line 94?